reconcile/fleet_labeler/integration.py

```diff
@@ -0,0 +1,151 @@
+        )
+        dependencies.populate_all()
+        self.reconcile(dependencies=dependencies)
+
+    def reconcile(self, dependencies: Dependencies) -> None:
+        validate_label_specs(specs=dependencies.label_specs_by_name)
+        if not dependencies.vcs:
+            raise ValueError("VCS is not initialized")
+        for spec_name, ocm in dependencies.ocm_clients_by_label_spec_name.items():
+            self._sync_cluster_inventory(
+                ocm, dependencies.label_specs_by_name[spec_name], dependencies.vcs
+            )
+
+    def _render_default_labels(
+        self,
+        template: FleetSubscriptionLabelTemplateV1,
+        labels: dict[str, str],
+    ) -> dict[str, Any]:
+        if not template.path:
+            # Make mypy happy
+            raise ValueError("path is required for subscription label template")
+        body = template.path.content
+        type = template.q_type or "jinja2"
+        extra_curly = type == "extracurlyjinja2"
+        vars = dict(template.variables or {})
+        vars["labels"] = labels
+        rendered = process_jinja2_template(
+            body,
+            vars,
+            extra_curly=extra_curly,
+        )
+        return yaml.safe_load(rendered)
+
+    def _render_yaml_file(
+        self,
+        current_content: str,
+        ids_to_delete: Iterable[str],
+        clusters_to_add: Iterable[YamlCluster],
+    ) -> str:
+        yml = create_ruamel_instance(pure=True)
+        content = yml.load(current_content)
+        current_clusters = content.get("clusters", [])
+        desired_clusters = [
+            cluster
+            for cluster in current_clusters
+            if cluster.get("clusterId") not in ids_to_delete
+        ]
+        desired_clusters.extend([
+            {
+                "name": cluster.name,
+                "clusterId": cluster.cluster_id,
+                "serverUrl": cluster.server_url,
+                "subscriptionLabels": cluster.subscription_labels_content,
+            }
+            for cluster in clusters_to_add
+        ])
+        content["clusters"] = desired_clusters
+        with StringIO() as stream:
+            yml.dump(content, stream)
+            return stream.getvalue()
+
+    def _sync_cluster_inventory(
+        self, ocm: OCMClient, spec: FleetLabelsSpecV1, vcs: VCS
+    ) -> None:
+        all_current_cluster_ids = [cluster.cluster_id for cluster in spec.clusters]
+        all_desired_cluster_ids: list[str] = []
+        clusters_to_add: list[YamlCluster] = []
+        for label_default in spec.label_defaults:
+            discovered_clusters_by_id = {
+                cluster.cluster_id: cluster
+                for cluster in ocm.discover_clusters_by_label_keys(
+                    keys=list(dict(label_default.match_subscription_labels).keys())
+                )
+                if dict(label_default.match_subscription_labels).items()
+                <= cluster.subscription_labels.items()
```

Comment:
what is comparing here dict items `<=` ?

Reply:
This compares whether match_subscription_labels is a subset of the clusters subscription_labels

Reply:
can it be done in server side label filter? or it's always `OR` condition match in server and we have to do another client side filter?

Reply:
Ah subscription_filter actually has an `eq` method too! Yeah will use that!

Reply:
Tried the `eq` filter, but it throws a 400 on OCM side. The or filter works well, so keeping it for now. Will try to figure out how to enhance this with eq filter for server side filtering.